File: internal/ai/routing/availability.lua
local action, token = ARGV[1], ARGV[2]
local lease = tonumber(ARGV[3])
local scopes, verdict = cjson.decode(ARGV[4]), cjson.decode(ARGV[5])
local clock = redis.call('TIME')
local now = tonumber(clock[1])*1000 + math.floor(tonumber(clock[2])/1000)
local prefix = 'dai:availability:v2:'
local permitKey = prefix..'permit:'..token
local ttl = 7*24*60*60*1000

local function load(key, scope)
  local raw = redis.call('GET', key)
  local s
  if raw then s = cjson.decode(raw)
  else s = {scope=scope, phase='available', epoch=0, trips=0, consecutive_failures=0,
    recovery_successes=0, recent_trips={}, window={}, stable_successes=0} end
  if s.phase == 'cooling' and (s.retry_at or 0) <= now then s.phase='recovering' end
  if (s.lease_until or 0) <= now then s.owner=nil; s.lease_until=0 end
  local trips = {}
  for _, t in ipairs(s.recent_trips or {}) do if t > now-3600000 then table.insert(trips,t) end end
  s.recent_trips=trips
  return s
end
local function save(key, s)
  redis.call('SET',key,cjson.encode(s),'PX',math.max(ttl,(s.retry_at or 0)-now+86400000))
  redis.call('SADD',prefix..'scopes',key)
  redis.call('SADD',prefix..'resource:'..s.scope.resource_kind..':'..s.scope.resource_id,key)
end
local function snapshot(s)
  local r = {busy=s.owner~=nil}
  for k,v in pairs(s) do if k ~= 'owner' and k ~= 'window' then r[k]=v end end
  -- Redis cjson represents an empty Lua table as an object. Keep public arrays stable.
  if #s.recent_trips == 0 then r.recent_trips=cjson.empty_array or nil end
  return r
end
local function array(items)
  if #items == 0 then return '[]' end
  return cjson.encode(items)
end
local function observe(s, failed)
  local sec = math.floor(now/1000)
  local total, failures, kept = 0,0,{}
  for k,b in pairs(s.window or {}) do
    if tonumber(k)>sec-60 then kept[k]=b; total=total+b[1]; failures=failures+b[2] end
  end
  local b = kept[tostring(sec)] or {0,0}
  b[1]=b[1]+1; if failed then b[2]=b[2]+1 end
  kept[tostring(sec)]=b; s.window=kept
  return total+1, failures+(failed and 1 or 0)
end
local function cool(s)
  s.trips=s.trips+1; s.epoch=s.epoch+1
  local delay=math.min(300000,30000*2^math.min(s.trips-1,4))
  if verdict.authentication then delay=1800000 end
  s.retry_at=now+delay
  if (verdict.cooldown_until or 0)>now and not verdict.authentication then s.retry_at=verdict.cooldown_until end
  s.phase='cooling'; s.owner=nil; s.lease_until=0; s.recovery_successes=0
  s.stable_at=nil; s.stable_successes=0
  table.insert(s.recent_trips,now)
end

if action=='cancel' then
 redis.call('SET',permitKey..':canceled','1','PX',math.max(lease,60000));action='complete';verdict={}
end
if action == 'complete' then
  local raw = redis.call('GET',permitKey)
  if not raw then return '[]' end
  local p = cjson.decode(raw)
  redis.call('DEL',permitKey)
  if p.expires_at<=now then return '[]' end
  local out={}
  for _, scope in ipairs(p.scopes) do
    local key=prefix..'scope:'..scope.key
    local s=load(key,scope)
    if s.epoch==p.epochs[scope.key] then
      -- A recovering lease reclaimed by another request fences this completion.
      local owns = s.phase~='recovering' or s.owner==token
      if owns then
        if s.owner==token then s.owner=nil; s.lease_until=0 end
        if verdict.success then
          s.verified_at=now; s.consecutive_failures=0
          if s.phase=='recovering' then
            s.recovery_successes=s.recovery_successes+1
            if s.recovery_successes>=2 then
              s.phase='available'; s.retry_at=0; s.window={}; s.stable_at=now; s.stable_successes=0
            end
          else
            observe(s,false)
            s.stable_at=s.stable_at or now; s.stable_successes=(s.stable_successes or 0)+1
            if now-s.stable_at>=600000 and s.stable_successes>=10 then s.trips=0 end
          end
        elseif verdict.failure_scope==scope.key then
          s.last_failure_at=now; s.reason=verdict.reason; s.stable_at=nil; s.stable_successes=0
          if not verdict.health_failure then cool(s)
          else
            s.consecutive_failures=s.consecutive_failures+1
            local n,f=observe(s,true)
            if s.phase=='recovering' or s.consecutive_failures>=5 or (n>=20 and f/n>=0.5) then cool(s) end
          end
        end
        save(key,s); table.insert(out,snapshot(s))
      end
    end
  end
  return array(out)
end

local states={}
for i,key in ipairs(KEYS) do states[i]=load(key,scopes[i]) end
if action=='import' or action=='suspend' then
 for i,s in ipairs(states) do
  if action=='suspend' or redis.call('EXISTS',KEYS[i])==0 then
   s.phase='cooling';s.epoch=s.epoch+1;s.retry_at=verdict.cooldown_until;s.reason=verdict.reason
   s.owner=nil;s.lease_until=0;s.recovery_successes=0;s.last_failure_at=now
   save(KEYS[i],s)
  end
 end
 return '[]'
end
if action=='read' then
  local out={}; for _,s in ipairs(states) do table.insert(out,snapshot(s)) end
  return array(out)
end
if action=='resume' or action=='reset' then
  for i,s in ipairs(states) do
   if action=='reset' or s.phase=='cooling' then
    s.epoch=s.epoch+1; s.phase='recovering'; s.retry_at=0; s.owner=nil; s.lease_until=0
    s.recovery_successes=0; s.window={}; s.consecutive_failures=0; s.stable_at=nil; s.stable_successes=0
    if action=='reset' then s.verified_at=nil end
    save(KEYS[i],s)
   end
  end
  return '[]'
end
if action=='acquire' then
 if redis.call('EXISTS',permitKey..':canceled')==1 then return cjson.encode({retry_at=0,reason='canceled'}) end
  local deniedAt,reason=0,''
  for _,s in ipairs(states) do
    if s.phase=='cooling' and s.retry_at>deniedAt then deniedAt=s.retry_at;reason=s.reason or 'cooling' end
    if s.phase=='recovering' and s.owner and s.lease_until>deniedAt then deniedAt=s.lease_until;reason='recovery_busy' end
  end
  if deniedAt>0 then return cjson.encode({retry_at=deniedAt,reason=reason}) end
  local p={token=token,epochs={},expires_at=now+lease,scopes=scopes}
  for i,s in ipairs(states) do
    p.epochs[s.scope.key]=s.epoch
    if s.phase=='recovering' then s.owner=token; s.lease_until=p.expires_at end
    save(KEYS[i],s)
  end
  redis.call('SET',permitKey,cjson.encode(p),'PX',lease)
  return cjson.encode({permit=p})
end
return redis.error_reply('unknown availability operation')

File: internal/ai/routing/availability.go
package routing

import (
	"context"
	_ "embed"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"sort"
	"time"

	"github.com/google/uuid"
	"github.com/redis/go-redis/v9"
	"xiaodou/dai/internal/ai/domain"
)

type AvailabilityPhase string

const (
	Available  AvailabilityPhase = "available"
	Cooling    AvailabilityPhase = "cooling"
	Recovering AvailabilityPhase = "recovering"
)

var ErrAvailabilityUnavailable = errors.New("availability state service unavailable")

// FaultScope identifies the smallest physical resource affected by a verdict.
// A group is deliberately absent: every tenant observes the same upstream.
type FaultScope struct {
	Key          string `json:"key"`
	Kind         string `json:"kind"`
	ResourceKind string `json:"resource_kind"`
	ResourceID   string `json:"resource_id"`
	EndpointID   string `json:"endpoint_id,omitempty"`
	CredentialID string `json:"credential_id,omitempty"`
	Model        string `json:"model,omitempty"`
	Operation    string `json:"operation,omitempty"`
}

func NewFaultScope(kind, resourceKind, resourceID, endpointID, credentialID, model, operation string) FaultScope {
	s := FaultScope{Kind: kind, ResourceKind: resourceKind, ResourceID: resourceID, EndpointID: endpointID, CredentialID: credentialID, Model: model, Operation: operation}
	b, _ := json.Marshal([]string{kind, resourceKind, resourceID, endpointID, credentialID, model, operation})
	s.Key = base64.RawURLEncoding.EncodeToString(b)
	return s
}

type AvailabilitySnapshot struct {
	Busy                bool              `json:"busy"`
	Scope               FaultScope        `json:"scope"`
	Phase               AvailabilityPhase `json:"phase"`
	Epoch               int64             `json:"epoch"`
	RetryAt             int64             `json:"retry_at,omitempty"`
	LeaseUntil          int64             `json:"lease_until,omitempty"`
	VerifiedAt          int64             `json:"verified_at,omitempty"`
	LastFailureAt       int64             `json:"last_failure_at,omitempty"`
	Reason              string            `json:"reason,omitempty"`
	ConsecutiveFailures int               `json:"consecutive_failures"`
	RecoverySuccesses   int               `json:"recovery_successes"`
	Trips               int               `json:"trips"`
	RecentTrips         []int64           `json:"recent_trips"`
}

func (s AvailabilitySnapshot) Blocked(now time.Time) bool {
	return s.Phase == Cooling && s.RetryAt > now.UnixMilli() || s.Phase == Recovering && s.Busy
}

type AdmissionPermit struct {
	Token     string           `json:"token"`
	Epochs    map[string]int64 `json:"epochs"`
	ExpiresAt int64            `json:"expires_at"`
}

type AdmissionDenied struct {
	RetryAt int64
	Reason  string
}

func (e *AdmissionDenied) Error() string { return "upstream temporarily unavailable: " + e.Reason }

// AvailabilityOutcome is independent of HTTP and billing. A neutral result
// always releases its own permit without asserting upstream health.
type AvailabilityOutcome struct {
	Success        bool   `json:"success"`
	FailureScope   string `json:"failure_scope,omitempty"`
	Reason         string `json:"reason,omitempty"`
	HealthFailure  bool   `json:"health_failure"`
	CooldownUntil  int64  `json:"cooldown_until,omitempty"`
	Authentication bool   `json:"authentication"`
}

type Availability interface {
	Read(context.Context, []FaultScope) (map[string]AvailabilitySnapshot, error)
	Acquire(context.Context, []FaultScope, time.Duration) (*AdmissionPermit, error)
	Complete(context.Context, *AdmissionPermit, AvailabilityOutcome) ([]AvailabilitySnapshot, error)
	List(context.Context, string, string) ([]AvailabilitySnapshot, error)
	Resume(context.Context, string, string) error
	RecoveryTurn(context.Context, string, []string, bool) (string, error)
}

//go:embed availability.lua
var availabilityLua string
var availabilityScript = redis.NewScript(availabilityLua)

// RedisAvailability runs the entire multi-scope transition atomically. Redis
// TIME, rather than a gateway's wall clock, controls cooldown and lease expiry.
type RedisAvailability struct{ client *redis.Client }

func NewRedisAvailability(client *redis.Client) *RedisAvailability {
	return &RedisAvailability{client: client}
}

func availabilityKey(scope string) string  { return "dai:availability:v2:scope:" + scope }
func resourceIndex(kind, id string) string { return "dai:availability:v2:resource:" + kind + ":" + id }

func (r *RedisAvailability) run(ctx context.Context, action string, scopes []FaultScope, token string, lease time.Duration, outcome AvailabilityOutcome) ([]byte, error) {
	if r == nil || r.client == nil {
		return nil, ErrAvailabilityUnavailable
	}
	ctx, cancel := context.WithTimeout(ctx, time.Second)
	defer cancel()
	keys := make([]string, len(scopes))
	for i, s := range scopes {
		keys[i] = availabilityKey(s.Key)
	}
	meta, _ := json.Marshal(scopes)
	verdict, _ := json.Marshal(outcome)
	value, err := availabilityScript.Run(ctx, r.client, keys, action, token, lease.Milliseconds(), string(meta), string(verdict)).Text()
	if err != nil {
		return nil, fmt.Errorf("%w: %v", ErrAvailabilityUnavailable, err)
	}
	return []byte(value), nil
}

func (r *RedisAvailability) Read(ctx context.Context, scopes []FaultScope) (map[string]AvailabilitySnapshot, error) {
	out := make(map[string]AvailabilitySnapshot, len(scopes))
	if len(scopes) == 0 {
		return out, nil
	}
	b, err := r.run(ctx, "read", scopes, "", 0, AvailabilityOutcome{})
	if err != nil {
		return nil, err
	}
	var records []AvailabilitySnapshot
	if err = json.Unmarshal(b, &records); err != nil {
		return nil, err
	}
	for _, rec := range records {
		out[rec.Scope.Key] = rec
	}
	return out, nil
}

func (r *RedisAvailability) Acquire(ctx context.Context, scopes []FaultScope, lease time.Duration) (*AdmissionPermit, error) {
	if lease <= 0 {
		lease = 17 * time.Minute
	}
	token := uuid.NewString()
	b, err := r.run(ctx, "acquire", scopes, token, lease, AvailabilityOutcome{})
	if err != nil {
		// A lost acknowledgement may hide a successful claim. The cancellation
		// tombstone also fences a delayed acquire arriving after this cleanup.
		cleanup, cancel := context.WithTimeout(context.Background(), time.Second)
		defer cancel()
		_, _ = r.run(cleanup, "cancel", nil, token, lease, AvailabilityOutcome{})
		return nil, err
	}
	var result struct {
		Permit  *AdmissionPermit `json:"permit"`
		RetryAt int64            `json:"retry_at"`
		Reason  string           `json:"reason"`
	}
	if err = json.Unmarshal(b, &result); err != nil {
		return nil, err
	}
	if result.Permit == nil {
		return nil, &AdmissionDenied{RetryAt: result.RetryAt, Reason: result.Reason}
	}
	return result.Permit, nil
}

func (r *RedisAvailability) Complete(ctx context.Context, permit *AdmissionPermit, result AvailabilityOutcome) ([]AvailabilitySnapshot, error) {
	if permit == nil {
		return nil, nil
	}
	b, err := r.run(ctx, "complete", nil, permit.Token, 0, result)
	if err != nil {
		return nil, err
	}
	var records []AvailabilitySnapshot
	err = json.Unmarshal(b, &records)
	return records, err
}

func (r *RedisAvailability) List(ctx context.Context, kind, id string) ([]AvailabilitySnapshot, error) {
	if r == nil || r.client == nil {
		return nil, ErrAvailabilityUnavailable
	}
	index := "dai:availability:v2:scopes"
	if kind != "" && id != "" {
		index = resourceIndex(kind, id)
	}
	keys, err := r.client.SMembers(ctx, index).Result()
	if err != nil {
		return nil, fmt.Errorf("%w: %v", ErrAvailabilityUnavailable, err)
	}
	if len(keys) == 0 {
		return []AvailabilitySnapshot{}, nil
	}
	pipe := r.client.Pipeline()
	cmds := make([]*redis.StringCmd, len(keys))
	for i, key := range keys {
		cmds[i] = pipe.Get(ctx, key)
	}
	_, err = pipe.Exec(ctx)
	if err != nil && err != redis.Nil {
		return nil, fmt.Errorf("%w: %v", ErrAvailabilityUnavailable, err)
	}
	scopes := make([]FaultScope, 0, len(keys))
	for _, c := range cmds {
		b, e := c.Bytes()
		if e == redis.Nil {
			continue
		}
		if e != nil {
			return nil, e
		}
		var rec AvailabilitySnapshot
		if e = json.Unmarshal(b, &rec); e != nil {
			return nil, e
		}
		scopes = append(scopes, rec.Scope)
	}
	states, err := r.Read(ctx, scopes)
	if err != nil {
		return nil, err
	}
	out := make([]AvailabilitySnapshot, 0, len(states))
	for _, rec := range states {
		out = append(out, rec)
	}
	sort.Slice(out, func(i, j int) bool { return out[i].Scope.Key < out[j].Scope.Key })
	return out, nil
}

func (r *RedisAvailability) Resume(ctx context.Context, kind, id string) error {
	states, err := r.List(ctx, kind, id)
	if err != nil {
		return err
	}
	scopes := make([]FaultScope, 0, len(states))
	for _, s := range states {
		scopes = append(scopes, s.Scope)
	}
	if len(scopes) == 0 {
		return nil
	}
	_, err = r.run(ctx, "resume", scopes, "", 0, AvailabilityOutcome{})
	return err
}

var recoveryTurnScript = redis.NewScript(`
local n = redis.call('INCR', KEYS[1]); redis.call('EXPIRE', KEYS[1], 3600)
if ARGV[1] == '1' and (n-1)%10 ~= 0 then return -1 end
local p = redis.call('INCR', KEYS[2]); redis.call('EXPIRE', KEYS[2], 3600)
return (p-1)%tonumber(ARGV[2])`)

func (r *RedisAvailability) RecoveryTurn(ctx context.Context, tier string, candidates []string, hasAvailable bool) (string, error) {
	if len(candidates) == 0 {
		return "", nil
	}
	if r == nil || r.client == nil {
		return "", ErrAvailabilityUnavailable
	}
	sort.Strings(candidates)
	flag := 0
	if hasAvailable {
		flag = 1
	}
	n, err := recoveryTurnScript.Run(ctx, r.client, []string{"dai:availability:v2:turn:" + tier, "dai:availability:v2:cursor:" + tier}, flag, len(candidates)).Int()
	if err != nil {
		return "", fmt.Errorf("%w: %v", ErrAvailabilityUnavailable, err)
	}
	if n < 0 {
		return "", nil
	}
	return candidates[n], nil
}

// Reset starts a new generation for only the configuration scopes that changed.
func (r *RedisAvailability) Reset(ctx context.Context, scopes []FaultScope) error {
	if len(scopes) == 0 {
		return nil
	}
	_, err := r.run(ctx, "reset", scopes, "", 0, AvailabilityOutcome{})
	return err
}

// ImportCooldown is idempotent and used only by the maintenance cutover.
func (r *RedisAvailability) ImportCooldown(ctx context.Context, scope FaultScope, until time.Time, reason string) error {
	_, err := r.run(ctx, "import", []FaultScope{scope}, "", 0, AvailabilityOutcome{CooldownUntil: until.UnixMilli(), Reason: reason})
	return err
}

// Suspend records explicit credential rejection from token refresh/diagnostics.
func (r *RedisAvailability) Suspend(ctx context.Context, scope FaultScope, until time.Time, reason string) error {
	_, err := r.run(ctx, "suspend", []FaultScope{scope}, "", 0, AvailabilityOutcome{CooldownUntil: until.UnixMilli(), Reason: reason})
	return err
}

// CandidateScopes is shared by credential selection and execution admission.
func CandidateScopes(c *domain.RouteCandidate, cred *domain.OAuthCredential) []FaultScope {
	if c == nil {
		return nil
	}
	kind, resource, endpoint, credential := "direct_upstream", c.EffectiveAccountID(), c.EndpointID, ""
	if c.IsPoolRoute() {
		kind, resource, endpoint = "oauth_pool", c.PoolID, ""
		if cred != nil {
			credential = cred.ID
		}
	}
	scopes := []FaultScope{NewFaultScope("transport", kind, resource, endpoint, "", "", "")}
	if !c.IsPoolRoute() || cred != nil {
		scopes = append(scopes, NewFaultScope("authentication", kind, resource, "", credential, "", ""), NewFaultScope("model", kind, resource, endpoint, credential, c.EffectiveUpstreamModel(), c.OperationKey()), NewFaultScope("rate_limit", kind, resource, endpoint, credential, c.EffectiveUpstreamModel(), c.OperationKey()))
	}
	return scopes
}
